Product category: Support updating existing category

itsmeichigo opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #10634 
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This PR adds support for updating existing product categories. Changes include:
- Added a new swipe action on the product category list for editing categories.
- Updated the title for the delete action to show text instead of icon for clarity.
- Updated `ProductCategoryListViewModel` to support finding `ProductCategory` models based on ID.
- Renamed `AddProductCategoryViewController` to `AddEditProductCategoryViewController` to support both adding and editing categories.
- Added a new view model for `AddEditProductCategoryViewController` and updated the controller to support editing categories.
- Updated `ProductParentCategoriesViewController` to display selected item and add option to remove current parent.

## Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->
- Log in to the app and navigate to the Products tab.
- Select an existing product or create a new one if needed.
- Select the category row on the product form or select Add more details > Category if the row is not already available.
- If there are no existing categories in your store, add at least one.
- On the category list, swipe left on any row. You should see a button to edit the item. Notice in Xcode console:  `🔵 Tracked product_category_settings_edit_button_tapped`.
- Tap the button, notice a new screen for editing category is displayed. The category name and parent should be correct.
- Edit any detail of the category. The Save button should only be enabled if there's any change to the category.
- Tap Save. The edit view should be dismissed and the category list should be updated.

Please feel free to check Menu > Coupon > Edit coupon > Select categories. Editing categories should not be enabled on this screen.


## Screenshots
<!-- Include before and after images or gifs when appropriate. -->


https://github.com/woocommerce/woocommerce-ios/assets/5533851/93aba02d-a46c-4b57-b40b-c05e137cf2cd




---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

itsmeichigo committed:
Add new action for editing product categories

itsmeichigo committed:
Rename AddEditProductCategoryViewController

itsmeichigo committed:
Add new view model for adding and editing category

itsmeichigo committed:
Rename xib file for AddEditProductCategoryViewController

itsmeichigo committed:
Update AddEditProductCategoryViewController with new view model

itsmeichigo committed:
Present edit category view when tapping edit category

itsmeichigo committed:
Update correct parent for edited category

itsmeichigo committed:
Track edit button tapped

itsmeichigo committed:
Display checkmark for selected parent category

itsmeichigo committed:
Add unit tests for AddEditProductCategoryViewModel

itsmeichigo committed:
Update release notes

itsmeichigo committed:
Add button to remove parent category

itsmeichigo committed:
Filter product parent category list to not show the child as an option

itsmeichigo committed:
Update remove parent button

itsmeichigo committed:
Merge branch 'feat/10634-delete-product-categories' into feat/10634-edit-product-categories

selanthiraiyan reviewed (commented):
https://github.com/woocommerce/woocommerce-ios/assets/524475/a2fe3053-c3b9-4e81-9060-74f15200e34f



Updating a category consistently fails for a JN site with the following error. 
```json
{
  "error": "internal_server_error",
  "message": "<p>There has been a critical error on this website.</p><p><a href=\"https://wordpress.org/documentation/article/faq-troubleshooting/\">Learn more about troubleshooting WordPress.</a></p>"
}
```


https://github.com/woocommerce/woocommerce-ios/assets/524475/fbc85b43-8219-4e7a-a49c-1ee79cb84f8b


Adding a category works for the same site, though. 🤔

itsmeichigo committed:
Update error alert title based on editing mode

itsmeichigo committed:
Update right bar button item for edit product category screen after failure

itsmeichigo committed:
Save only category name and parent ID when updating product categories

itsmeichigo commented:
Thanks @selanthiraiyan for the reviews!

> Updating a category consistently fails for a JN site with the following error.

Good catch! Looks like this is an issue with self-hosted sites only - I updated the remote to only send category name and parent ID when updating categories since the issue might have happened due to the unexpected `siteID` sent together with other category details. I also fixed issues with the UI in the failure case. This PR is now ready for another look.

itsmeichigo committed:
Merge branch 'feat/10634-delete-product-categories' into feat/10634-edit-product-categories

itsmeichigo committed:
Update release notes

selanthiraiyan reviewed (approved):
This works as expected. 🎉
